File: frontend/src/app/features/directory/services/directory.service.ts
import { Injectable, inject } from '@angular/core';
import { HttpClient, HttpParams } from '@angular/common/http';
import { Observable } from 'rxjs';
import { environment } from '../../../../environments/environment';

/**
 * Alumni card data for directory listing.
 * Contains essential information needed for directory grid display.
 * Used in alumni directory listings and search results.
 */
export interface AlumniCard {
  /** Unique identifier for the alumnus */
  id: string;
  /** First name of the alumnus */
  firstName: string;
  /** Last name of the alumnus */
  lastName: string;
  /** URL to the alumnus's profile image */
  profileImageUrl?: string;
  /** Year the alumnus graduated from IHCAE */
  graduationYear?: number;
  /** Batch/Year the alumnus graduated from IHCAE (Backend DTO compatibility) */
  batch?: string;
  /** Course/program the alumnus completed at IHCAE */
  course?: string;
  /** Current job title or position */
  jobTitle?: string;
  /** Current location (city, state, country) */
  location?: string;
  /** Email address of the alumnus */
  email: string;
}

/**
 * Detailed alumni profile data for individual profile pages.
 * Contains comprehensive information including contact details.
 * Used for the alumni detail page with quick action buttons.
 */
export interface AlumniDetail {
  /** Unique identifier for the alumnus */
  id: string;
  /** First name of the alumnus */
  firstName: string;
  /** Last name of the alumnus */
  lastName: string;
  /** Email address for contact purposes */
  email: string;
  /** Phone number for contact purposes */
  phone?: string;
  /** URL to the alumnus's profile image */
  profileImageUrl?: string;
  /** Year the alumnus graduated from IHCAE */
  graduationYear?: number;
  /** Batch/Year the alumnus graduated from IHCAE (Backend DTO compatibility) */
  batch?: string;
  /** Course/program the alumnus completed at IHCAE */
  course?: string;
  /** Personal biography or description */
  bio?: string;
  /** Current job title or position */
  jobTitle?: string;
  /** Current location (city, state, country) */
  location?: string;
  /** Date when the alumnus joined the platform */
  createdAt: string;
}

/**
 * Paginated result wrapper for API responses.
 * Contains the data items and pagination metadata.
 */
export interface PaginatedResult<T> {
  /** Array of items for the current page */
  items: T[];
  /** Total number of items across all pages */
  totalCount: number;
  /** Current page number (1-based) */
  page: number;
  /** Number of items per page */
  pageSize: number;
}

/**
 * Directory filter parameters for search and filtering.
 * Used to construct query parameters for the directory API.
 */
export interface DirectoryFilters {
  /** Optional search term for name or email filtering */
  search?: string;
  /** Optional course filter (e.g., "Advanced Mountaineering") */
  course?: string;
  /** Optional graduation year filter (e.g., 2020) */
  graduationYear?: number;
  /** Optional batch filter */
  batch?: string;
  /** Page number (1-based, default: 1) */
  page?: number;
  /** Number of records per page (default: 20) */
  pageSize?: number;
}

/**
 * Service for alumni directory operations.
 * Handles directory listing, search, filtering, and alumni detail retrieval.
 * Provides methods for browsing alumni and accessing detailed profile information.
 * 
 * Features:
 * - Paginated alumni directory with search and filtering
 * - Detailed alumni profile retrieval
 * - Convenience methods for filter options
 * - Error handling and type safety
 * 
 * @author IHCAE Development Team
 * @version 1.0.0
 */
@Injectable({
  providedIn: 'root'
})
export class DirectoryService {
  private http = inject(HttpClient);
  private apiUrl = `${environment.apiUrl}/api/v1/alumni`;

  /**
   * Gets a paginated, filtered list of approved alumni for the directory.
   * Supports search by name/email and filtering by course/graduation year.
   * Only returns approved alumni members for security.
   * 
   * @param filters Optional filters for search, course, graduation year, and pagination
   * @returns Observable of paginated alumni cards
   */
  getAlumniDirectory(filters: DirectoryFilters = {}): Observable<PaginatedResult<AlumniCard>> {
    // Build query parameters from filters
    let params = new HttpParams();

    if (filters.search) {
      params = params.set('search', filters.search);
    }
    if (filters.course) {
      params = params.set('course', filters.course);
    }
    if (filters.graduationYear) {
      params = params.set('batch', filters.graduationYear.toString());
    }
    if (filters.batch) {
      params = params.set('batch', filters.batch);
    }
    if (filters.page) {
      params = params.set('page', filters.page.toString());
    }
    if (filters.pageSize) {
      params = params.set('pageSize', filters.pageSize.toString());
    }

    // Make HTTP GET request to alumni directory endpoint
    return this.http.get<PaginatedResult<AlumniCard>>(this.apiUrl, { params });
  }

  /**
   * Gets detailed information for a specific alumnus.
   * Returns comprehensive profile data including contact information.
   * Used for the alumni detail page with quick action buttons.
   * 
   * @param userId The unique identifier of the alumnus
   * @returns Observable of detailed alumni information
   */
  getAlumniDetail(userId: string): Observable<AlumniDetail> {
    return this.http.get<AlumniDetail>(`${this.apiUrl}/${userId}`);
  }

  /**
   * Convenience method to get all available courses from the directory.
   * Extracts unique courses by querying a large page of results.
   */
  getAvailableCourses(): Observable<string[]> {
    return new Observable(observer => {
      // Fetch a large page to get all courses
      this.getAlumniDirectory({ pageSize: 100 }).subscribe({
        next: (result) => {
          // Extract unique courses from results
          const courses = new Set<string>();
          result.items.forEach(alumni => {
            if (alumni.course) {
              courses.add(alumni.course);
            }
          });
          // Return sorted array of unique courses
          observer.next(Array.from(courses).sort());
          observer.complete();
        },
        error: (error) => observer.error(error)
      });
    });
  }

  /**
   * Convenience method to get graduation years available in the directory.
   * Extracts unique graduation years by querying a large page of results.
   */
  getAvailableYears(): Observable<number[]> {
    return new Observable(observer => {
      // Fetch a large page to get all graduation years
      this.getAlumniDirectory({ pageSize: 100 }).subscribe({
        next: (result) => {
          // Extract unique graduation years from results
          const years = new Set<number>();
          result.items.forEach(alumni => {
            if (alumni.graduationYear) {
              years.add(alumni.graduationYear);
            }
          });
          // Return sorted array of unique years (newest first)
          observer.next(Array.from(years).sort((a, b) => b - a));
          observer.complete();
        },
        error: (error) => observer.error(error)
      });
    });
  }

  /**
   * Convenience method to get distinct batch values available in the directory.
   * Extracts unique batch strings by querying a large page of results.
   */
  getAvailableBatches(): Observable<string[]> {
    return new Observable(observer => {
      this.getAlumniDirectory({ pageSize: 100 }).subscribe({
        next: (result) => {
          const batches = new Set<string>();
          result.items.forEach(alumni => {
            if (alumni.batch) {
              batches.add(alumni.batch);
            }
          });
          // Return sorted array of unique batches (newest first)
          observer.next(Array.from(batches).sort().reverse());
          observer.complete();
        },
        error: (error) => observer.error(error)
      });
    });
  }
}

File: backend/IHCAE.Api/Features/Directory/Services/DirectoryService.cs
using IHCAE.Api.Features.Directory.Models.DTOs;
using IHCAE.Api.Features.Auth.Repositories;
using IHCAE.Api.Features.Auth.Models.Entities;
using IHCAE.Api.Shared.DTOs;
using IHCAE.Api.Shared.Constants;
using IHCAE.Api.Shared.Services;

namespace IHCAE.Api.Features.Directory.Services;

/// <summary>
/// Service for managing alumni directory operations.
/// Handles business logic for browsing, searching, and viewing alumni profiles.
/// Provides data transformation between entities and DTOs for API responses.
/// Implements security by only returning approved alumni members.
/// </summary>
public class DirectoryService : IDirectoryService
{
    private readonly IUserRepository _userRepository;
    private readonly ILogger<DirectoryService> _logger;
    private readonly IUrlHelperService _urlHelperService;

    /// <summary>
    /// Initializes the DirectoryService with required dependencies.
    /// </summary>
    /// <param name="userRepository">Repository for User entity operations</param>
    /// <param name="logger">Logger for tracking directory operations</param>
    /// <param name="urlHelperService">Service for converting relative URLs to absolute URLs</param>
    public DirectoryService(
        IUserRepository userRepository,
        ILogger<DirectoryService> logger,
        IUrlHelperService urlHelperService)
    {
        _userRepository = userRepository;
        _logger = logger;
        _urlHelperService = urlHelperService;
    }

    /// <summary>
    /// Gets a paginated, filtered list of approved alumni for the directory.
    /// Delegates to UserRepository for data access and applies business logic.
    /// Transforms User entities to AlumniCardDto objects for API responses.
    /// </summary>
    /// <param name="searchTerm">Optional search term for name or email filtering</param>
    /// <param name="course">Optional course filter (e.g., "Advanced Mountaineering")</param>
    /// <param name="graduationYear">Optional graduation year filter (e.g., 2020)</param>
    /// <param name="page">Page number (1-based indexing)</param>
    /// <param name="pageSize">Number of records per page (default: 20, max: 100)</param>
    /// <returns>Paginated result containing AlumniCardDto objects and metadata</returns>
    /// <exception cref="ArgumentException">Thrown when page or pageSize parameters are invalid</exception>
    public async Task<PaginatedResult<AlumniCardDto>> GetAlumniDirectoryAsync(
        string? searchTerm = null,
        string? course = null,
        string? batch = null,
        int page = 1,
        int pageSize = 20)
    {
        // Validate pagination parameters
        if (page < 1)
        {
            throw new ArgumentException("Page number must be greater than 0", nameof(page));
        }
        
        if (pageSize < 1 || pageSize > 100)
        {
            throw new ArgumentException("Page size must be between 1 and 100", nameof(pageSize));
        }

        // Retrieve filtered alumni data from repository
        var (users, totalCount) = await _userRepository.GetApprovedAlumniAsync(
            searchTerm, course, batch, page, pageSize);

        // Transform User entities to AlumniCardDto objects
        var alumniCards = users.Select(MapToAlumniCardDto).ToList();

        _logger.LogInformation("Retrieved {Count} alumni for directory page {Page}", alumniCards.Count, page);

        // Return paginated result with metadata
        return new PaginatedResult<AlumniCardDto>
        {
            Items = alumniCards,
            TotalCount = totalCount,
            Page = page,
            PageSize = pageSize
        };
    }

    /// <summary>
    /// Gets detailed information for a specific alumnus.
    /// Validates that the user exists and is an approved alumni member.
    /// Returns comprehensive profile data including contact information.
    /// </summary>
    /// <param name="userId">The unique identifier of the alumnus</param>
    /// <returns>Detailed alumni information including contact details</returns>
    /// <exception cref="KeyNotFoundException">Thrown when alumnus is not found</exception>
    /// <exception cref="UnauthorizedAccessException">Thrown when user is not approved</exception>
    public async Task<AlumniDetailDto> GetAlumniDetailAsync(Guid userId)
    {
        // Retrieve user with profile data
        var user = await _userRepository.GetWithProfileAsync(userId);

        if (user == null)
        {
            throw new KeyNotFoundException($"Alumni with ID {userId} not found");
        }

        // Ensure user is an approved alumni member with "Alumni" role
        if (user.Status != UserStatus.Approved)
        {
            throw new UnauthorizedAccessException("User is not an approved alumni member");
        }

        // Check if user has "Alumni" role (exclude administrators and other roles)
        if (!user.UserRoles.Any(ur => ur.Role.Name == RoleConstants.Alumni))
        {
            throw new UnauthorizedAccessException("User is not an alumni member");
        }

        _logger.LogInformation("Retrieved detailed profile for alumni {UserId}", userId);

        // Transform User entity to AlumniDetailDto
        return MapToAlumniDetailDto(user);
    }

    /// <summary>
    /// Maps a User entity to an AlumniCardDto for directory card display.
    /// Extracts essential information needed for directory grid view.
    /// Handles null AlumniProfile gracefully by providing default values.
    /// </summary>
    /// <param name="user">The User entity to map</param>
    /// <returns>AlumniCardDto containing card display information</returns>
    private AlumniCardDto MapToAlumniCardDto(User user)
    {
        return new AlumniCardDto
        {
            Id = user.Id,
            FirstName = user.FirstName,
            LastName = user.LastName,
            ProfileImageUrl = _urlHelperService.GetAbsoluteUrl(user.AlumniProfile?.ProfileImageUrl),
            Batch = user.AlumniProfile?.Batch,
            Course = user.AlumniProfile?.Course,
            JobTitle = user.AlumniProfile?.JobTitle,
            Location = user.AlumniProfile?.Location
        };
    }

    /// <summary>
    /// Maps a User entity to an AlumniDetailDto for detailed profile view.
    /// Includes all available information including contact details.
    /// Handles null AlumniProfile gracefully by providing default values.
    /// </summary>
    /// <param name="user">The User entity to map</param>
    /// <returns>AlumniDetailDto containing complete profile information</returns>
    private AlumniDetailDto MapToAlumniDetailDto(User user)
    {
        return new AlumniDetailDto
        {
            Id = user.Id,
            FirstName = user.FirstName,
            LastName = user.LastName,
            Email = user.Email,
            Phone = user.Phone,
            ProfileImageUrl = _urlHelperService.GetAbsoluteUrl(user.AlumniProfile?.ProfileImageUrl),
            Batch = user.AlumniProfile?.Batch,
            Course = user.AlumniProfile?.Course,
            Bio = user.AlumniProfile?.Bio,
            JobTitle = user.AlumniProfile?.JobTitle,
            Location = user.AlumniProfile?.Location,
            CreatedAt = user.CreatedAt
        };
    }
}

